### PR TITLE
chore: moved span names processing to a tracer component

### LINKED ISSUE #564: Centralized component for span processing
Span names for some of the endpoints, that are handled before auth checks, like "/version" and "/health", are not formatted well, their names consist of HTTP method only. There could be introduced a new centralized component, which is responsible for formatting/processing span names before the actual processing of incoming requests.

### PR DESCRIPTION
**Pull request by alekseyvdovenko**
Closes https://github.com/epam/ai-dial-core/issues/564

### CODE REVIEW

**Comment by alekseyvdovenko** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-565.nightly-test.deltixhub.io
>E2E tests status: success

File: server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelection.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package com.epam.aidial.core.server.controller;
+
+import com.epam.aidial.core.server.Proxy;
+import com.epam.aidial.core.server.ProxyContext;
+
+import java.util.function.BiFunction;
+
+public final class ControllerSelection {
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelection.java]: ControllerTemplate?

File: server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelection.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    private final String pathTemplate;
+    private final BiFunction<Proxy, ProxyContext, Controller> controllerProvider;
+
+    ControllerSelection(String pathTemplate, BiFunction<Proxy, ProxyContext, Controller> controllerProvider) {
+        this.pathTemplate = pathTemplate;
+        this.controllerProvider = controllerProvider;
+    }
+
+    public Controller controller(Proxy proxy, ProxyContext context) {
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelection.java]: controller -> build

File: server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java [hunk 3]
```diff
@@ -339,57 +271,76 @@ private static Controller selectPost(Proxy proxy, ProxyContext context, String p
...
-        return null;
+    public ControllerSelection select(String path) {
+        return select(null, path);
+    }
+
+    private ControllerSelection select(HttpMethod method, String path) {
+        return ROUTES.stream()
+                .map(r -> r.select(method, path))
+                .filter(Objects::nonNull)
+                .max(Comparator.comparing(ControllerSelection::pathTemplate))
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java]: Why do we need to call max? Perhaps just findAny()?

File: server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java [hunk 4]
```diff
@@ -405,4 +356,22 @@ private String resourcePath(String url) {
 
         return url.substring(prefix.length());
     }
-}
\ No newline at end of file
+
+    private record ControllerRoute(HttpMethod method, Pattern pathPattern, Initializer initializer) {
+        public ControllerSelection select(HttpMethod method, String path) {
+            if (method == null || this.method.equals(method)) {
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java]: `if (method == null || this.method.equals(method))` -> `if (this.method.equals(method))`

File: server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java [hunk 5]
```diff
@@ -17,6 +23,13 @@
 @UtilityClass
 public class ControllerSelector {
 
+    private static final Object CTX_KEY = new Object();
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java]: CTX -> CONTROLLER_TEMPLATE

File: server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java [hunk 6]
```diff
@@ -3,16 +3,23 @@
...
+import java.util.Set;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 @UtilityClass
 public class RegexUtil {
 
-    public String replaceNamedGroups(Pattern pattern, String input, List<String> groups) {
-        if (groups == null || groups.isEmpty()) {
+    public String replaceNamedGroups(Pattern pattern, String input) {
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java]: input -> path?

File: server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java [hunk 7]
```diff
@@ -49,6 +56,17 @@ private List<RegexGroup> collectGroups(Pattern pattern, String input, List<Strin
         return regexGroups;
     }
 
+    public static Set<String> getNamedGroups(Pattern pattern) {
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java]: private?

File: server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java [hunk 8]
```diff
@@ -3,16 +3,23 @@
...
 public class RegexUtil {
 
-    public String replaceNamedGroups(Pattern pattern, String input, List<String> groups) {
-        if (groups == null || groups.isEmpty()) {
+    public String replaceNamedGroups(Pattern pattern, String input) {
+        return replaceNamedGroups(pattern, input, getNamedGroups(pattern));
+    }
+
+    public String replaceNamedGroups(Pattern pattern, String input, Collection<String> groups) {
+        if (pattern == null || input == null || input.isBlank() || groups == null || groups.isEmpty()) {
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java]: We should check the first
`if (pattern == null || input == null || input.isBlank()` before making a list of groups

File: server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java (hunk 6 above)
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java]: We can keep just one method as of now

File: server/src/main/java/com/epam/aidial/core/server/tracing/OpenTelemetryTracingFactoryProxy.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package com.epam.aidial.core.server.tracing;
+
+import io.vertx.core.spi.VertxTracerFactory;
+import io.vertx.core.spi.tracing.VertxTracer;
+import io.vertx.core.tracing.TracingOptions;
+import io.vertx.tracing.opentelemetry.OpenTelemetryTracingFactory;
+
+public class OpenTelemetryTracingFactoryProxy extends OpenTelemetryTracingFactory {
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/tracing/OpenTelemetryTracingFactoryProxy.java]: OpenTelemetryTracingFactoryProxy -> DialTracingFactory?

File: server/src/main/java/com/epam/aidial/core/server/tracing/SpanNameTracer.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+import io.vertx.core.spi.tracing.SpanKind;
+import io.vertx.core.spi.tracing.TagExtractor;
+import io.vertx.core.spi.tracing.VertxTracer;
+import io.vertx.core.tracing.TracingPolicy;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+public class SpanNameTracer<I, O> implements VertxTracer<I, O> {
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/tracing/SpanNameTracer.java]: SpanNameTracer -> DialVertxTracer?

File: server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java [hunk 11]
```diff
@@ -3,16 +3,23 @@
...
+import java.util.Collection;
 import java.util.Comparator;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 @UtilityClass
 public class RegexUtil {
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java:14]: Can we move the class to ControllerSelector and change access level to package ?

**Review by astsiapanay — CHANGES_REQUESTED** [review]

File: server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelection.java (hunk 1 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelection.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelection.java (hunk 2 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelection.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java (hunk 5 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java (hunk 6 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java (hunk 7 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java (hunk 8 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java (hunk 6 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/tracing/OpenTelemetryTracingFactoryProxy.java (hunk 9 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/tracing/OpenTelemetryTracingFactoryProxy.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/tracing/SpanNameTracer.java (hunk 10 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/tracing/SpanNameTracer.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java (hunk 11 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/util/RegexUtil.java:14]: fixed

File: server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java (hunk 3 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java (hunk 4 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java]: fixed

**Review by astsiapanay — DISMISSED** [review]

File: server/src/main/java/com/epam/aidial/core/server/tracing/DialVertxTracer.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        HttpMethod method = request.method();
+        String path = request.path();
+
+        if (HttpMethod.GET.equals(method) && PATHS.contains(path)) {
+            return "%s %s".formatted(method, path);
+        }
+        if (HttpMethod.OPTIONS.equals(method)) {
+            return method.name();
+        }
+        ControllerTemplate selection = ControllerSelector.select(request);
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/tracing/DialVertxTracer.java:75]: Could be null? Add a test that sends a request with path /fake.

Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/tracing/DialVertxTracer.java:75]: No, by default `RouteController` is selected when we cannot match (This is how it works now and this logic remains the same). Anyway, added test cases

Comment (artsiomkorzun) [reply server/src/main/java/com/epam/aidial/core/server/tracing/DialVertxTracer.java:75]: Ok

File: server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java [hunk 13]
```diff
@@ -339,57 +270,72 @@ private static Controller selectPost(Proxy proxy, ProxyContext context, String p
...
             return () -> controller.handle(resourcePath(path));
-        }
+        });
+    }
 
-        return null;
+    public ControllerTemplate select(HttpServerRequest request) {
+        HttpMethod method = request.method();
+        String path = request.path();
+        if (request instanceof HttpServerRequestInternal req) {
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java:306]: Where do you use it?

Comment (artsiomkorzun) [reply server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java:306]: Seems like you just want to cache the result in tracer and the following code that selects the controller.

**Review by astsiapanay — APPROVED** [review]

**Pull request merged**